Port macOS and Linux release packages to .NET 5

pchote opened the pull request:
This follows #17989 by switching over the macOS packages, which requires two additional changes to work well:
1. Assemblies are now installed into the `mod.app/Contents/MacOS` subdir, using `Engine.EngineDir` to point to the data in `mod.app/Contents/Resources`. This brings us in line with the app bundling convention for putting all executable code in `MacOS`, and more importantly fixes the app icon otherwise being missing.
2. Use [hard links](https://en.wikipedia.org/wiki/Hard_link) within the disk image to avoid duplicating the otherwise identical runtime files. This reduced the final dmg size from 143MB to 84MB, compared to 60MB for the current mono based image.

I could only find [one reference](https://stackoverflow.com/questions/11940188/is-there-a-way-to-merge-redundant-files-when-preparing-a-dmg-file) online to anybody using this hard link trick before, so i'm not 100% sure this will be compatible on all systems. It works for me on macOS 11.0.1, but we should get some additional testing on at least macOS 10.13 before we merge this.

Test builds: https://github.com/pchote/OpenRA/releases/tag/devtest-20201223-4

pchote committed:
Replace duplicate runtime files with hardlinks to reduce dmg size.

pchote committed:
Trim unused assemblies to reduce packaged size further.

pchote committed:
Use ULFO format for non-compat disk images (requires macOS 10.11+).

pchote commented:
Added a commit to enable trimming, which saves a further 10MB on both macOS and Windows, and another commit to switch the non-compat builds to the ULFO format for faster decompression.

Updated test builds: https://github.com/pchote/OpenRA/releases/tag/devtest-20201223-8

Windows should now be tested too since it is affected by trimming.

pchote commented:
N/A confirmed on Discord that the -8 test build works on macOS 10.13.

pchote committed:
Run appimagetool directly.

pchote committed:
Remove obsolete --check-runtime-assemblies utility command.

pchote commented:
Rebased on latest version of #17989 and added some more commits to switch the Linux AppImages to NET 5 and remove the now-obsolete runtime assemblies check. This now just leaves the makefile and related bits that were dropped out of the first PR to complete the full migration.

Updated test builds: https://github.com/pchote/OpenRA/releases/tag/devtest-20201228

I have tested that the AppImages run fine on Ubuntu 14.04 and CentOS 7, so we should not be dropping any distros with this switch.

pchote commented:
* Need to update the "Used By" comments
* Need to change the mono installation to copy all dlls to compensate for the removed assemblies check.

teinarss commented on OpenRA.Launcher/OpenRA.Launcher.csproj:
```diff
@@ -65,4 +65,8 @@
   </Target>
+  <ItemGroup>
+    <TrimmerRootAssembly Include="netstandard" />
+    <TrimmerRootAssembly Include="System.IO.Pipes" />
```
Do we need this line? Ie do we have anything dependent on pipes?

pchote replied:
Apparently so, since we crash without this... I don't know why the trimmer doesn't see this though, maybe referenced by one of our thirdparty deps loaded via a mod dll?

teinarss replied:
What is the crash message? Also i cant find any dependency on pipes in our 3d party deps.

pchote replied:
```
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner
  Exception of type `System.IO.FileNotFoundException`: Could not load file or assembly 'System.IO.Pipes, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

     at DiscordRPC.IO.ManagedNamedPipeClient..ctor()
   at DiscordRPC.DiscordRpcClient..ctor(String applicationID, Int32 pipe, ILogger logger, Boolean autoEvents, INamedPipeClient client)
   at OpenRA.Mods.Common.DiscordService..ctor(MiniYaml yaml) in /Users/paul/src/OpenRA/OpenRA.Mods.Common/DiscordService.cs:line 62
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at OpenRA.Manifest.LoadCustomData(ObjectCreator oc) in /Users/paul/src/OpenRA/OpenRA.Game/Manifest.cs:line 176
   at OpenRA.ModData..ctor(Manifest mod, InstalledMods mods, Boolean useLoadScreen) in /Users/paul/src/OpenRA/OpenRA.Game/ModData.cs:line 60
   at OpenRA.Game.InitializeMod(String mod, Arguments args) in /Users/paul/src/OpenRA/OpenRA.Game/Game.cs:line 446
   at OpenRA.Game.Initialize(Arguments args) in /Users/paul/src/OpenRA/OpenRA.Game/Game.cs:line 409
   at OpenRA.Game.InitializeAndRun(String[] args) in /Users/paul/src/OpenRA/OpenRA.Game/Game.cs:line 270
   at OpenRA.Launcher.Program.Main(String[] args) in /Users/paul/src/OpenRA/OpenRA.Launcher/Program.cs:line 30
```
Makes sense that this would be using a pipe.

teinarss replied:
Then this line would only to be in the csproj that reference discordprc, right?

pchote replied:
Doesn't work. As far as I can tell `TrimmerRootAssembly` is ignored on anything except `<OutputType>Exe</OutputType>`

pchote committed:
Switch macOS packages to .NET 5.

pchote committed:
Switch Appimages to net 5.

pchote committed:
Change install_assemblies_mono to install all bin contents.

pchote commented:
Updated. I'll wait for #17989 to be merged before tagging another test build.

pchote commented:
Rebased.

Updated (and hopefully final) test builds: https://github.com/pchote/OpenRA/releases/tag/devtest-20210101
Packaging logs: https://github.com/pchote/OpenRA/actions/runs/456542700

Mailaender commented:
The `OpenRA-Dune-2000-devel-x86_64.AppImage` immediately crashes in skirmish and after a while also on the shellmap:

```
Exception of type `System.TypeInitializationException`: The type initializer for 'OpenRA.Mods.Common.Traits.BotModules.Squads.AttackOrFleeFuzzy' threw an exception.
Inner
  Exception of type `System.IO.FileNotFoundException`: Could not load file or assembly 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.

     at OpenRA.Mods.Common.Traits.BotModules.Squads.AttackOrFleeFuzzy..ctor(IEnumerable`1 rulesForNormalOwnHealth, IEnumerable`1 rulesForInjuredOwnHealth, IEnumerable`1 rulesForNeadDeadOwnHealth)
   at OpenRA.Mods.Common.Traits.BotModules.Squads.AttackOrFleeFuzzy..cctor() in /home/runner/work/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/BotModules/Squads/AttackOrFleeFuzzy.cs:line 98
   at OpenRA.Mods.Common.Traits.BotModules.Squads.GroundUnitsIdleState.Tick(Squad owner) in /home/runner/work/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs:line 54
   at OpenRA.Mods.Common.Traits.SquadManagerBotModule.AssignRolesToIdleUnits(IBot bot) in /home/runner/work/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:line 235
   at OpenRA.Mods.Common.Traits.SquadManagerBotModule.OpenRA.Mods.Common.Traits.IBotTick.BotTick(IBot bot) in /home/runner/work/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:line 182
   at OpenRA.Mods.Common.Traits.ModularBot.<OpenRA.Traits.ITick.Tick>b__14_0() in /home/runner/work/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/Player/ModularBot.cs:line 94
   at OpenRA.Sync.<>c__DisplayClass13_0.<RunUnsynced>b__0() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Sync.cs:line 166
   at OpenRA.Sync.RunUnsynced[T](Boolean checkSyncHash, World world, Func`1 fn) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Sync.cs:line 190
   at OpenRA.Sync.RunUnsynced(Boolean checkSyncHash, World world, Action fn) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Sync.cs:line 167
   at OpenRA.Mods.Common.Traits.ModularBot.OpenRA.Traits.ITick.Tick(Actor self) in /home/runner/work/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/Player/ModularBot.cs:line 90
   at OpenRA.TraitDictionary.TraitContainer`1.ApplyToAllTimed(Action`2 action, String text) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/TraitDictionary.cs:line 307
   at OpenRA.World.Tick() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/World.cs:line 435
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 646
   at OpenRA.Game.LogicTick() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 661
   at OpenRA.Game.Loop() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 823
   at OpenRA.Game.Run() in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 864
   at OpenRA.Game.InitializeAndRun(String[] args) in /home/runner/work/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 274
   at OpenRA.Launcher.Program.Main(String[] args) in /home/runner/work/OpenRA/OpenRA/OpenRA.Launcher/Program.cs:line 30
```

```
Platform is Linux
Engine version is devtest-20210101
Runtime: .NET CLR 5.0.1
```

pchote commented:
Fixed.

New test build: https://github.com/pchote/OpenRA/releases/tag/devtest-20210102

Mailaender reviewed (approved):
Tested on Linux.

pchote commented:
Is there anything else I can do here to help get this PR merged?

I wanted to delay the .NET 5 migration into Next + 1 so we could maximize the amount of testing and time to work around any issues on the new toolchain. This time is currently going wasted, until the next PR (which depends on this one) is merged.

abcdefg30 reviewed (approved):
Portable build worked fine on Windows 10.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/74b167fc02bd7d871755a75359822eb4d1030ac9...5979493359ffafbbad8732565e00a2b8195f7b90)
